src/lib/transaction_pool/transaction_pool.ml

```diff
@@ -317,68 +276,162 @@ struct
+                      Option.bind
+                        (Base_ledger.location_of_key ledger
+                           (User_command.sender tx))
+                        ~f:(Base_ledger.get ledger)
                     with
-                    | Ok () ->
-                        Logger.debug t.logger ~module_:__MODULE__
-                          ~location:__LOC__
-                          ~metadata:
-                            [ ( "user_cmd"
-                              , User_command.to_yojson (txn :> User_command.t)
-                              ) ]
-                          "Adding $user_cmd to my pool locally, and \
-                           scheduling for rebroadcast" ;
-                        (add' pool txn, (txn :> User_command.t) :: acc)
-                    | Error err ->
-                        Logger.faulty_peer t.logger ~module_:__MODULE__
-                          ~location:__LOC__
-                          !"Got transaction not valid against strongest \
-                            ledger: %{sexp: \
-                            User_command.With_valid_signature.t}, because \
-                            %{sexp: Error.t} rejecting."
-                          txn err ;
-                        (pool, acc) ) ) )
-      in
-      t.pool <- pool' ;
-      match res with
-      | [] ->
-          Deferred.Or_error.error_string "No new transactions"
-      | xs ->
-          Deferred.Or_error.return xs
+                    | None ->
+                        let%bind _ =
+                          trust_record
+                            ( Trust_system.Actions.Sent_useless_gossip
+                            , Some
+                                ( "account does not exist for command: $cmd"
+                                , [("cmd", User_command.to_yojson tx)] ) )
+                        in
+                        go txs'' pool accepted
+                    | Some account ->
+                        let sufficient_fee =
+                          match Indexed_pool.min_fee pool with
+                          | None ->
+                              true
+                          | Some min_fee ->
+                              if Indexed_pool.size pool >= pool_max_size then
+                                Currency.Fee.(User_command.fee tx > min_fee)
+                              else true
+                        in
+                        if sufficient_fee then
+                          let res, pool' =
+                            Indexed_pool.add_from_gossip_exn pool tx'
+                              account.nonce
+                            @@ Currency.Balance.to_amount account.balance
+                          in
+                          let yojson_fail_reason =
+                            Fn.compose
+                              (fun s -> `String s)
+                              (function
+                                | `Success _ ->
+                                    failwith "impossible"
```

Comment:
making this polymorphic variant into a result would make it so you don't need this failwith here which is nice

Reply:
Should we be using `failwith`s for unreachable situations, in general?